### ISSUE TITLE
[REQ] SmartSpin2K & QZ integration

### ISSUE THREAD

**Issue by marklogan17** [cagnulein/qdomyos-zwift#299]
Looking to integrate the two so we can remotely control resistance on bikes with no motorized resistance control.  NRF screenshots attached.  I have started a discussion on the SmartSpin2K github page asking for suggestions/thoughts.
![Screenshot_20210505-181402_nRF Connect](https://user-images.githubusercontent.com/75054215/117223300-617da780-addb-11eb-99b1-8727df343fcd.jpg)
![Screenshot_20210505-181437_nRF Connect](https://user-images.githubusercontent.com/75054215/117223302-62163e00-addb-11eb-9d49-747f847c2441.jpg)
![Screenshot_20210505-181420_nRF Connect](https://user-images.githubusercontent.com/75054215/117223303-62aed480-addb-11eb-9a0d-3ac9a9042b6e.jpg)
![Screenshot_20210505-181446_nRF Connect](https://user-images.githubusercontent.com/75054215/117223304-63476b00-addb-11eb-9881-34eebb931988.jpg)
![Screenshot_20210505-181352_nRF Connect](https://user-images.githubusercontent.com/75054215/117223305-63476b00-addb-11eb-89d1-e10a5a513a8e.jpg)
![Screenshot_20210505-181511_nRF Connect](https://user-images.githubusercontent.com/75054215/117223306-63476b00-addb-11eb-93a4-45b584c86895.jpg)

**Comment by cagnulein** [comment]
Ok nice, it's a standard ftms.
So let's talk about the implementation: i guess i should treat this as an
accessory for every bike, not as a bike itself.
Do you agree?
So new settings to enable it and configure it

Il gio 6 mag 2021, 01:52 marklogan17 ***@***.***> ha scritto:

**Comment by marklogan17** [comment]
Yes I think its an accessory for every bike.  My concern is what feedback SmartSpin will need from the bike to know when its changed the resistance?  In other words, does SmartSpin need to be told the current resistance level on the bike so it knows when to stop turning the knob?  I think the answer is 'no' but I don't know that :) 

I can get reasonable logging on the web interface to SmartSpin so even though I don't have the motor I should be able to see it 'trying' to change the resistance if it receives the commands from QZ.

**Comment by cagnulein** [comment]
mmm interesting thought. But SmartSpin should works even with bike without
bluetooth, so i don't think it needs a feedback. It just counts its motor
steps.

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno gio 6 mag 2021 alle ore 14:10 marklogan17 <
***@***.***> ha scritto:

**Comment by marklogan17** [comment]
Yes that's what I'm hoping, but is that accurate for every resistance level?  i.e. Does changing from 24 to 25 resistance have the same amount of turning as changing from 31 to 32 resistance?  I 'hope' its consistent :)

**Comment by cagnulein** [comment]
it has to be! i don't want to write the smartspin firmware also :D

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno gio 6 mag 2021 alle ore 14:19 marklogan17 <
***@***.***> ha scritto:

**Comment by cagnulein** [comment]
give it a try @marklogan17 
there is a new parameter called "FTMS accessory".
If i did everything right, when zwift or peloton asks for a resistance change, i will send also to SpinBike2K :)

**Comment by marklogan17** [comment]
Built it.  It connected to the smartspin (going to have to ask them for a switch to turn on/off echelon support so smartspin doesn't grab connection to bike).  I tried a few times but didn't get any messages to come across to the smart spin. Put Zwift setting to Erg mode just for the fun of it.  Tried changing resistance on target resistance and regular resistance tiles.  Turned on/off the 'auto follow resistance' at the top of the app.  Some log files attached.

[debug-Mon_May_10_13_36_01_2021.log](https://github.com/cagnulein/qdomyos-zwift/files/6454212/debug-Mon_May_10_13_36_01_2021.log)
[debug-Mon_May_10_13_37_59_2021.log](https://github.com/cagnulein/qdomyos-zwift/files/6454213/debug-Mon_May_10_13_37_59_2021.log)

**Comment by kadaan** [comment]
@cagnulein We only support two modes for FTMS Control at the moment: 0x05 and 0x11

**Comment by cagnulein** [comment]
> @cagnulein We only support two modes for FTMS Control at the moment: 0x05 and 0x11

Yes i saw. I already fork your project and i would like to add also the resistance command to.
What do you think?

In any case i think i can emulate it send a grade command proportionally to the resistance requested

**Comment by kadaan** [comment]
"To the resistance required" will be difficult without first "homing" the knob to zero.

**Comment by cagnulein** [comment]
> Fork and contribute would be preferred.  I'd love to see if there is a way to collaborate for the benefit of all.

Yes of course, i will do a PR when i will have something ready :)

**Comment by cagnulein** [comment]
> "To the resistance required" will be difficult without first "homing" the knob to zero.

It's not the same with ZWIFT also? I mean how did you homing with the current version? It's something requested to the user?

**Comment by kadaan** [comment]
With zwift, in sim mode, it controls the incline.  The user can set the resistance via SS2Ks remote switches on the handlebars.  All control from Zwift is relative to that.  For ERG mode we track the target range and vary resistance to try to match that. 

For Sim and ERG, you can just signal SS2K via Sim.  If Zwift is signaling you with sim mode control you can pass it though.  If Zwift is controlling with ERG, you have target and current power, so you calculate the needed change and send it as incline.

Target resistance mode is the only one that need homing.  

We want to support homing, but need to determine the best way to do it.  Different bike have different means of controlling resistance and I'm not sure that they will all trigger the stepper motors end stop.  We only have a few bikes to test on, so it is a small sample set.

**Comment by kadaan** [comment]
Yeah.  That is the plan, and it will work for bikes with mechanical control of the resistance (ie. IC4).  Bike with a different type of control (ie. electronic) that will not work.

**Comment by cagnulein** [comment]
It seems that you already know better than me how the world works :D
I guess manual homing is the safest choice so

Il mar 11 mag 2021, 07:18 Joel Baranick ***@***.***> ha
scritto:

**Comment by kadaan** [comment]
On a side note, I'm trying to build your iOs app and this keeps happening:
```
Undefined symbols for architecture arm64:
  "_QZ_EnableDiscoveryCharsAndDescripttors", referenced from:
      echelonconnectsport::echelonconnectsport(bool, bool, unsigned char, double) in echelonconnectsport.o
      fitplusbike::fitplusbike(bool, bool, unsigned char, double) in fitplusbike.o
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

Any ideas?

**Comment by cagnulein** [comment]
> On a side note, I'm trying to build your iOs app and this keeps happening:
> 
> ```
> Undefined symbols for architecture arm64:
>   "_QZ_EnableDiscoveryCharsAndDescripttors", referenced from:
>       echelonconnectsport::echelonconnectsport(bool, bool, unsigned char, double) in echelonconnectsport.o
>       fitplusbike::fitplusbike(bool, bool, unsigned char, double) in fitplusbike.o
> ld: symbol(s) not found for architecture arm64
> clang: error: linker command failed with exit code 1 (use -v to see invocation)
> ```
> 
> Any ideas?

yes because you have to use these https://github.com/cagnulein/qdomyos-zwift/tree/master/qt-patches/ios/5.15.2

**Comment by cagnulein** [comment]
> Hehehe. Having no app makes it harder for us :P

this is where QZ can help :)
P.S. where are you based?

**Comment by cagnulein** [comment]
@kadaan you can just use the binary library without touching the mm, this is why i added to the GIT repo (in order to help people to build QZ without rebuilding the QT)

**Comment by kadaan** [comment]
Where can I get the debug logs?  The app based on master keeps crashing on me.

**Comment by cagnulein** [comment]
ok we did some steps forward @marklogan17 . i guess you have now some issue about virtual device without QT patched. Try to disable the virtual device and retry please.

> With the FTMS field populated but the smartspin turned off, QZ actually failed to connect to the bike

did you wait at least 10 seconds in this scenario? i have a timeout about this condition.

Also in the log i saw not the smartspin but it fails everytime, so i hope it's releated to the virtual bike setting

**Comment by marklogan17** [comment]
Tried again.  Disabled virtual device and restarted.  With the smartspin OFF QZ did not connect to the bike - I waited about 15 or 20 seconds.  Disabled FTMS and it connected right away.  Started SmartSpin and I don't think I got a connection - tried a few things.  Finally restarted the SmartSpin and it looked like it connected however again no control messages came across.  
[debug-Tue_May_11_09_42_50_2021.log](https://github.com/cagnulein/qdomyos-zwift/files/6459648/debug-Tue_May_11_09_42_50_2021.log)

Next I tried QZ with the SmartSpin already running - had to be a little careful of timing because I didn't want the SmartSpin connecting to the Echelon directly.  This time QZ started, connected to bike and again looked like it connected to SmartSpin but no data.
[debug-Tue_May_11_09_47_27_2021.log](https://github.com/cagnulein/qdomyos-zwift/files/6459663/debug-Tue_May_11_09_47_27_2021.log)

**Comment by cagnulein** [comment]
it works! i mean not from your point of view, but in the log, in the 2nd log, i saw the connection between the bike and between the SS2K at the same time!

Now i have to fixed 2 bugs:
1) the first scenario that you described
2) the resistance issue (i already found the cause)

I will tell you when i will be ready to try again

**Comment by cagnulein** [comment]
i got the "15" error. try to lower resistance (lower than 15), it should change.
anyway i'm pushing it :)

**Comment by kadaan** [comment]
Any reason to have a limit of 100?

**Comment by cagnulein** [comment]
> Any reason to have a limit of 100?

just a workaround, i will read from the ftms in the next versions

**Comment by cagnulein** [comment]
> Also, it would be good to get the SS2K switches to work with QZ

one thing at a time :)

**Comment by marklogan17** [comment]
> Also, it would be good to get the SS2K switches to work with QZ

I thought about this too but since QZ reads from the bike, changing the level by the shifters will get echoed to QZ via the bike.  So it 'kinda' will...

**Comment by kadaan** [comment]
> > Also, it would be good to get the SS2K switches to work with QZ
> 
> I thought about this too but since QZ reads from the bike, changing the level by the shifters will get echoed to QZ via the bike. So it 'kinda' will...

Except AFAIK not all bikes expose resistance over FTMS

**Comment by marklogan17** [comment]
> > > Also, it would be good to get the SS2K switches to work with QZ
> > 
> > 
> > I thought about this too but since QZ reads from the bike, changing the level by the shifters will get echoed to QZ via the bike. So it 'kinda' will...
> 
> Except AFAIK not all bikes expose resistance over FTMS

True.  Some are proprietary (Echelon) and some have nothing.  Are you thinking of some web api that apps could subscribe to to be notified on internal smartspin changes?  You would almost need to disable the internal processing of the shifters if pushing control out to QZ or other apps.

**Comment by cagnulein** [comment]
You should expose shifts in the resistance attributes on ftms. What do you
think?

Il mar 11 mag 2021, 17:50 Joel Baranick ***@***.***> ha
scritto:

**Comment by marklogan17** [comment]
Great news!  I got my build working - at least to the point where the motor turns.  I can now use QZ to cause the motor to turn either direction by hitting the + or - button on the resistance tile.  I'll try and get the build buttoned up tomorrow so that I can actually install it on the bike and tweak the parameters to get it turning the right amount to change resistance levels.  Very exciting!

**Comment by cagnulein** [comment]
Can't wait to see it working on the bike!

Il mer 12 mag 2021, 02:14 marklogan17 ***@***.***> ha scritto:

**Comment by marklogan17** [comment]
I installed on the bike.  I had some binding issues that I got sorted out enough to test.  I set the # of steps to 750 which seems to equate to a level of resistance each time I hit a shifter button.  When I use the +/- buttons on QZ they send an incline value that is equal to the next resistance level.  So for example if I'm at resistance 22 and hit + it will send an incline of 23 if I hit '-' it sends an incline of 21.  The first time you do that on start up you tend to get a huge resistance change.  I went from 24 to 32 one time for example.  I also had a huge shift down the one time.  I'm assuming that's because SmartSpin is assuming some starting incline, probably zero, and when you immediately put '23' it spins a long way.  Also 1 increase of incline doesn't seem to be necessarily equivalent to 1 increase of resistance in smartspin, it feels like it spins less.  I don't have anything empirical to back that up.

So, in short, its working and its close to good enough.  If we tweak the incline stuff a little we'll probably have something really workable.  In the mean time, I'm continuing to figure out my binding issues.  I'm also going to see if I can build the smartspin fork that was provided that has more info coming across the bluetooth.

**Comment by kadaan** [comment]
As of right now the method @cagnulein is using to change SS2K resistance is setting "incline".  SS2K assumes you start at "incline" of zero, and all changes from that baseline.  In practice this works correctly for Zwift, FulGaz, etc. in sim mode because the "incline" changes are directly related virtual changes in slope.

There is an incline multiplier that can impact how much the knob is turned and is applied to the data that @cagnulein is sending.

This sounds like it will "work" for now, but is not optimal.  The data coming from @cagnulein sounds like absolute values, but the api being used expects relative values.  The shifters in SS2K change the baseline and all incline changes are relative to that.  That way you can adjust the "feel" in realtime if things are too tough or too easy.

The real solution here is to add an SS2K BLE service that:

1. Allows disabling of direct stepper control from within SS2K
2. Disables FTMS, HRM, CPS BLE binding
3. Notifies on shifter up and down press
4. Enables direct stepper control

With these capabilities added, QZ could take over the BLE communication with bikes and apps, if desired, and it could drive SS2K's control of the knob.

For users operating without QZ, SS2K would retain standard functionality.

**Comment by cagnulein** [comment]
@kadaan is there a way with the current firmware to have access to the current inclination/resistance of SS2K?
Or maybe for now i can set my homing point to the current resistance and put it as an offset. What do you think @marklogan17 ?

Of course having the raw control of the motor will be the perfect solution

**Comment by kadaan** [comment]
@cagnulein No way ATM to get access to current inclination/resistance
Don't think that will work because QZ and SS2K don't know what the resistance of the bike is.

**Comment by cagnulein** [comment]
@kadaan with the offset, SS2K will become just an actuators so it should work. I'm only worried about negative inclines value. It will work for sure with positive ones

**Comment by kadaan** [comment]
If by "set my homing point to the current resistance" you mean the user indicates in the app what resistance the bike is set to before starting a ride, then yes, that could work.

**Comment by cagnulein** [comment]
@kadaan yeah but without the user intervention: you know, i know already which is the current resistance. The bike told me it :)
So SS2K will work also with the negative incline?

**Comment by kadaan** [comment]
@cagnulein Yes, negative incline is fine.  But, not all bikes tell you the current resistance (ie. Schwinn IC4/IC8)

**Comment by cagnulein** [comment]
@kadaan not true :D we have an algorithm to calculate the resistance for the schwinn :) it will works with your bike too :)

**Comment by kadaan** [comment]
@cagnulein Cool.  I want to see how accurate it is.  Seems unlikely to match the display, but if it does, that's great!!!

**Comment by kadaan** [comment]
@cagnulein Gave it a very quick try.  Seems close when the resistance is at the higher end, 70+.  On my bike it is very inaccurate at lower resistances < 50.  When the bike is at 0, the resistance display in QZ is 28.

**Comment by cagnulein** [comment]
@marklogan17 i added the startup offset to the smartspin2k module. could you please try? if something doesn't work, as always, send me the debug log

**Comment by marklogan17** [comment]
OK you guys are smarter than me - plus you seem to solve things while I'm sleeping.  I know the long term solution is likely to disable the shifters and have the request pass through to QZ.  However until that's done I'm not sure how best to handle this.  I tried a bunch of things to get this consistently changing 1 resistance level every time a + or - button is hit on QZ however I seemed to run into various edge cases.  

What would be ideal is for a simple way of QZ replicating the shifters rather than dealing with the relative nature of the incline.  So it would send the number of 'shifter presses' that it needs to replicate.  That way the step config in SmartSpin still applies nicely.   Sending a 'number' of shifter presses matters because for Schwinn or Peleton or something like that you may need to increase by 10 or 20 resistance levels and you don't want to make 20 calls, waiting for the resistance change between each.

**Comment by kadaan** [comment]
@marklogan17 The issue is that "api" for controlling the stepper is not in terms of shifter steps, but in incline.  Incline is scaled by Incline Multiplier, so you will want to adjust that to try to get one shift per incline submitted by QZ.

**Comment by marklogan17** [comment]
> @marklogan17 The issue is that "api" for controlling the stepper is not in terms of shifter steps, but in incline. Incline is scaled by Incline Multiplier, so you will want to adjust that to try to get one shift per incline submitted by QZ.

OK, I was not reading the code right where it was taking current position and adding incline*incline multiplier.  You caused me to re-read it and now I think we're good.

I've settled on 800 as the shift steps that I need for each level.  So @cagnulein what we need is the inclination change you send times the configured multiplier equal 800.  So the easiest way is where you are multiplying by 100 and sending, multiply by 200 and then I set the incline multiplier to 4.  So 1 resistance level will cause you to send an incline change of 200 (or -200) which SmartSpin multiplies by 4, getting us to the +/-800 that seems to be good for changing one resistance level.  Even better, if this is applied to the Peleton usage, you can easily request changes greater than 1 resistance level and it should still work.

I modified QZ to multiply by 200 and did a quick test and it worked really well.  We'll still get potentially messed up in cases where the shifters are used and then +/- but I'll figure all those edge cases out so I can tell people what to do/avoid - and more importantly, how to get back in sync.

Longer term we probably need a way to reset inclination to zero so we can resync between apps.

Now to pull the hardware apart and replace the small gear to hopefully get rid of my last binding issue :)

**Comment by marklogan17** [comment]
@kadaan any update on the changes in SmartSpin?

**Comment by kadaan** [comment]
@marklogan17 No, I've been crazy busy on my farm. :(

**Comment by marklogan17** [comment]
> @marklogan17 No, I've been crazy busy on my farm. :(

Totally understandable!  Life comes first :)

**Comment by cagnulein** [comment]
#377

**Comment by cagnulein** [comment]
i'm closing this one. in case we will open a new one for future requests!
